### PR TITLE
feat: add TypeTransformer for Tensorflow Model

### PR DESCRIPTION
**Pull request by techytushar**
# TL;DR
This PR adds support for:
1. TypeTransformer for Tensorflow Model type

**Note:** I was not able to run the `pip-compile` commands to generate the new `requirements` files, it failed on `protobuf` version error

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
[https://github.com/flyteorg/flyte/issues/2570](https://github.com/flyteorg/flyte/issues/2570)

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by pingsutw** [comment]
> I was not able to run the pip-compile commands to generate the new requirements files, it failed on protobuf version error

Adding `grpcio-status<1.49.0` to the requirement.in can probably fix it.

File: flytekit/extras/tensorflow/__init__.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+"""
```
Comment (pingsutw) [inline flytekit/extras/tensorflow/__init__.py:1]: Could we move these files to `flytekit/extras/tensorflow/model`? we'll have two transformers (tensor and model) under the `tensorFlow` directory. https://github.com/flyteorg/flytekit/pull/1243

**Review by pingsutw — COMMENTED** [review]
Could we move these files to `flytekit/extras/tensorflow/model`?

**Review by pingsutw — COMMENTED** [review]
Two tests are failing. `ModuleNotFoundError: No module named 'tensorflow'`
You have to run `make dev-requirements.txt` to update the requirements in `dev-requirements.txt`, and then GA will be able to install TensorFlow.

File: flytekit/extras/tensorflow/model/native.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+            literal_type.blob is not None
+            and literal_type.blob.dimensionality == _core_types.BlobType.BlobDimensionality.SINGLE
+            and literal_type.blob.format == self.TENSORFLOW_FORMAT
+        ):
+            return T
+
+        raise ValueError(f"Transformer {self} cannot reverse {literal_type}")
+
+
+class TensorflowModelTransformer(TensorflowTypeTransformer[tf.keras.Model]):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/model/native.py:71]: @techytushar, `tf.Module` is the parent class for `tf.keras.Model` and `tf.keras.layers.Layer`. Can you use that here? Also, can you merge this class into the parent class cause I don't think we need to use Generic in the first place if there's only one override?

**Comment by samhita-alla** [comment]
@pingsutw, what do you think about having a `model.py` rather than a model folder, or better `module.py` since the type transformer will be for `tf.Module`? It seems to make more sense to me. Since we also have a PR for tensor, we can have a `tensor.py` file.

File: tests/flytekit/unit/extras/tensorflow/test_native.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+def generate_model() -> tf.keras.Model:
+    inputs = tf.keras.Input(shape=(32,))
+    outputs = tf.keras.layers.Dense(1)(inputs)
+    model = tf.keras.Model(inputs, outputs)
+    return model
+
+
+@task
+def get_model_layers(model: tf.keras.Model) -> List[tf.keras.layers.Layer]:
+    return model.layers
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/test_native.py:18]: This test shouldn't use your type transformer; it should just pickle. So it doesn't make sense to have unless you add a type transformer for `tf.Module`.

Comment (samhita-alla) [reply tests/flytekit/unit/extras/tensorflow/test_native.py:18]: Also, I'm unsure if List[...] will work. `tf.keras.layers.Layer` should, however, work after you add `tf.Module`.

File: tests/flytekit/unit/extras/tensorflow/test_native.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/test_native.py:20]: @techytushar, can you also add a test for `tf.keras.Sequential` and check if that's working?

**Comment by pingsutw** [comment]
>  do you think about having a model.py rather than a model folder
yes. make sense. I prefer `module.py`

File: flytekit/extras/tensorflow/__init__.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+.. autosummary::
+   :template: custom.rst
+   :toctree: generated/
+"""
+from flytekit.loggers import logger
+
+# TODO: abstract this out so that there's an established pattern for registering plugins
+# that have soft dependencies
+try:
+    # isolate the exception to the torch import
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:15]:
```suggestion
    # isolate the exception to the tensorflow import
```

File: flytekit/extras/tensorflow/__init__.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    import tensorflow
+
+    _tensorflow_installed = True
+except (ImportError, OSError):
+    _tensorflow_installed = False
+
+
+if _tensorflow_installed:
+    from .model import TensorflowModelTransformer
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:24]:
```suggestion
    from .module import TensorFlowModuleTransformer
```

Can you change the file name to `module.py` and the transformer name to `TensorFlowModuleTransformer`?

File: flytekit/extras/tensorflow/__init__.py [hunk 7]
```diff
@@ -0,0 +1,26 @@
+"""
+Flytekit Tensorflow
+=========================================
+.. currentmodule:: flytekit.extras.tensorflow
+
+.. autosummary::
+   :template: custom.rst
+   :toctree: generated/
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:8]: I don't think we need any documentation for this. Can you remove this docstring, delete `extras.tensorflow.rst` and remove the tensorflow entry in `types.extend.rst`?

File: flytekit/extras/tensorflow/model.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+import tensorflow as tf
+
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+
+
+class TensorflowModelTransformer(TypeTransformer[tf.keras.Model]):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/model.py:13]:
```suggestion
class TensorflowModelTransformer(TypeTransformer[tf.Module]):
```

Please check the comments we've added.

File: tests/flytekit/unit/extras/tensorflow/test_native.py (hunk 4 above)
Comment (techytushar) [reply tests/flytekit/unit/extras/tensorflow/test_native.py]: added

File: flytekit/extras/tensorflow/model.py (hunk 8 above)
Comment (techytushar) [reply flytekit/extras/tensorflow/model.py:13]: I tried using the `tf.Module` but the `tf.keras.models.save_model` method fails when saving the `Layers` (in the unit tests).
I don't think we can save the whole layer in tensorflow, we can only save its weights

How should I go about it ?
1. keep the type as `tf.keras.Model`
2. or use `tf.Module` and add separate check for `Layers` and save them separately

Comment (samhita-alla) [reply flytekit/extras/tensorflow/model.py:13]: @techytushar, you're right. How about then you create two files, one each for model and layers? You can name as `model.py` and `layer.py`, respectively. The model transformer also needs to subclass `tf.keras.Sequential` along with `tf.keras.Model`. Please fine the layer serialization [here](https://www.tensorflow.org/api_docs/python/tf/keras/layers/serialize).

Comment (samhita-alla) [reply flytekit/extras/tensorflow/model.py:13]: For layer one, set the literal to `generic`, e.g., https://github.com/flyteorg/flytekit/blob/6530601c2538a5d804127a97f63291730b1ba1d8/plugins/flytekit-dolt/flytekitplugins/dolt/schema.py#L76.

Comment (techytushar) [reply flytekit/extras/tensorflow/model.py:13]: The `Sequential` [class](https://www.tensorflow.org/api_docs/python/tf/keras/Sequential) already inherits from the `Model` class, so is the subclassing needed ? If it is a `Sequential` type then it automatically use the `ModelTransformer`

Comment (samhita-alla) [reply flytekit/extras/tensorflow/model.py:13]: Oh, then not needed.

**Comment by samhita-alla** [comment]
@techytushar, let me know once the changes are in.

**Comment by techytushar** [comment]
@samhita-alla I have pushed the changes.

**Note:** I could not use the generic Struct type for storing the layer config because in protobuf format the `int` values gets converted to `float` due to which when we convert it back to the `tf layer` it was not able to parse it. So I have stored them as simple `json string`

More info here: https://stackoverflow.com/questions/51818125/how-to-use-ints-in-a-protobuf-struct

File: flytekit/extras/tensorflow/__init__.py [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+    # isolate the exception to the tensorflow import
+    import tensorflow
+
+    _tensorflow_installed = True
+except (ImportError, OSError):
+    _tensorflow_installed = False
+
+
+if _tensorflow_installed:
+    from .layer import TensorflowLayerTransformer
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:21]: Can you please change the class names to `TensorFlow...` with `F` capitalized?

File: flytekit/extras/tensorflow/__init__.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+"""
+Flytekit Tensorflow
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:2]:
```suggestion
Flytekit TensorFlow
```

File: flytekit/extras/tensorflow/__init__.py [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+except (ImportError, OSError):
+    _tensorflow_installed = False
+
+
+if _tensorflow_installed:
+    from .layer import TensorflowLayerTransformer
+    from .model import TensorflowModelTransformer
+else:
+    logger.info(
+        "Unable to register TensorflowModelTransformer, TensorflowLayerTransformer because tensorflow is not installed."
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/__init__.py:25]:
```suggestion
        "Unable to register TensorFlowModelTransformer and TensorFlowLayerTransformer because tensorflow is not installed."
```

File: flytekit/extras/tensorflow/layer.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+from flytekit.core.context_manager import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer, TypeTransformerFailedError
+from flytekit.models.literals import Literal, Primitive, Scalar
+from flytekit.models.types import LiteralType, SimpleType
+
+
+class TensorflowLayerTransformer(TypeTransformer[tf.keras.layers.Layer]):
+    def __init__(
+        self,
+    ):
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/layer.py:15]:
```suggestion
    def __init__(self):
```

File: flytekit/extras/tensorflow/layer.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
+import json as _json
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/layer.py:1]: I don't think an alias is required here. Let's just import json: `import json`.

File: flytekit/extras/tensorflow/layer.py [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        self, ctx: FlyteContext, lv: Literal, expected_python_type: Type[tf.keras.layers.Layer]
+    ) -> tf.keras.layers.Layer:
+        if not (lv and lv.scalar and lv.scalar.primitive):
+            raise TypeTransformerFailedError(f"Cannot convert from {lv} to {expected_python_type}")
+
+        layer_config = _json.loads(lv.scalar.primitive.string_value)
+        return tf.keras.layers.deserialize(layer_config)
+
+    def guess_python_type(self, literal_type: LiteralType) -> Type[tf.keras.layers.Layer]:
+        if literal_type.simple == literal_type.simple.STRUCT:
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/layer.py:40]:
```suggestion
        if literal_type is not None and (literal_type.simple == SimpleType.STRING):
```

File: flytekit/extras/tensorflow/model.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+        expected: LiteralType,
+    ) -> Literal:
+        meta = BlobMetadata(
+            type=_core_types.BlobType(
+                format=self.TENSORFLOW_FORMAT,
+                dimensionality=_core_types.BlobType.BlobDimensionality.SINGLE,
+            )
+        )
+
+        local_path = ctx.file_access.get_random_local_path() + ".h5"
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/model.py:41]:
```suggestion
        local_path = ctx.file_access.get_random_local_path()
```

File: flytekit/extras/tensorflow/model.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        pathlib.Path(local_path).parent.mkdir(parents=True, exist_ok=True)
+
+        # save tensorflow model in h5 format
```
Comment (samhita-alla) [inline flytekit/extras/tensorflow/model.py:44]:
```suggestion
        # save tensorflow model in tf format
```

File: tests/flytekit/unit/extras/tensorflow/test_layer.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+        loss=tf.keras.losses.BinaryCrossentropy(),
+        metrics=[
+            tf.keras.metrics.BinaryAccuracy(),
+        ],
+    )
+    return model.layers
+
+
+@task
+def get_layers_weights(layers: List[tf.keras.layers.Layer]) -> List[Any]:
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/test_layer.py:34]: Shouldn't be `List[Any]`. If they are floats, use `List[float]`.

File: tests/flytekit/unit/extras/tensorflow/test_layer.py [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+
+@workflow
+def wf():
+    dense_layer = get_layer()
+    layers = generate_sequential_model()
+    get_layers_weights(layers=[dense_layer])
+    get_layers_weights(layers=layers)
+
+
+@workflow
```
Comment (samhita-alla) [inline tests/flytekit/unit/extras/tensorflow/test_layer.py:46]: Please remove this.

File: flytekit/extras/tensorflow/model.py (hunk 15 above)
Comment (techytushar) [reply flytekit/extras/tensorflow/model.py:41]: `tf.keras.models.save_model` supports [2 formats](https://www.tensorflow.org/guide/keras/save_and_serialize#whole-model_saving_loading), by default it is `SavedModel` format but it stores the model as a folder and not as a single file. Due to which this line fails
```
ctx.file_access.put_data(local_path, remote_path, is_multipart=False)
```
So I used the `h5` format since it stores it as a single file.

Comment (samhita-alla) [reply flytekit/extras/tensorflow/model.py]: You should be able to copy a directory as well by setting `is_multipart` to True.

**Comment by techytushar** [comment]
@samhita-alla since its the last day of Hacktoberfest, please can you merge this PR or add label so that it can count towards my contribution

**Comment by samhita-alla** [comment]
@techytushar, I'll merge the PR later as I need to push some changes.

**Comment by cosmicBboy** [comment]
fixes https://github.com/flyteorg/flyte/issues/2570.

We're going to need to move these changes into the tensorflow plugin @eapolinario

**Comment by eapolinario** [comment]
> fixes [flyteorg/flyte#2570](https://github.com/flyteorg/flyte/issues/2570).
> 
> We're going to need to move these changes into the tensorflow plugin @eapolinario

Great, I'll make sure to incorporate those there.
